shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/spi/ordered/cache/OrderedServicesCache.java

```diff
@@ -17,50 +17,45 @@
 
 package org.apache.shardingsphere.infra.spi.ordered.cache;
 
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
 import lombok.AccessLevel;
 import lombok.NoArgsConstructor;
 import org.apache.shardingsphere.infra.spi.ordered.OrderedSPI;
 
-import java.util.Collection;
 import java.util.Map;
 import java.util.Optional;
-import java.util.concurrent.ConcurrentHashMap;
 
 /**
  * Ordered services cached.
  */
 @NoArgsConstructor(access = AccessLevel.PRIVATE)
 public final class OrderedServicesCache {
     
-    private static final Map<Class<?>, CachedOrderedServices> CACHED_SERVICES = new ConcurrentHashMap<>();
+    private static final Cache<OrderedServicesCacheKey, Map<?, ?>> CACHED_SERVICES = CacheBuilder.newBuilder().softValues().build();
     
     /**
      * Find cached services.
      * 
-     * @param types types
-     * @param orderedSPIClass class of ordered SPI
+     * @param cacheKey cache key
      * @param <K> type of key
      * @param <V> type of ordered SPI class
      * @return cached ordered services
      */
-    public static <K, V extends OrderedSPI<?>> Optional<CachedOrderedServices> findCachedServices(final Collection<K> types, final Class<V> orderedSPIClass) {
-        return isHitCache(types, orderedSPIClass) ? Optional.of(CACHED_SERVICES.get(orderedSPIClass)) : Optional.empty();
-    }
-    
-    private static <K, V extends OrderedSPI<?>> boolean isHitCache(final Collection<K> types, final Class<V> orderedSPIClass) {
-        return CACHED_SERVICES.containsKey(orderedSPIClass) && CACHED_SERVICES.get(orderedSPIClass).getTypes().equals(types);
+    @SuppressWarnings("unchecked")
+    public static <K, V extends OrderedSPI<?>> Optional<Map<K, V>> findCachedServices(final OrderedServicesCacheKey cacheKey) {
+        return Optional.ofNullable(CACHED_SERVICES.getIfPresent(cacheKey)).map(optional -> (Map<K, V>) optional);
     }
     
     /**
      * Cache services.
      * 
-     * @param types types
-     * @param orderedSPIClass class of ordered SPI
+     * @param cacheKey cache key
      * @param services ordered services
      * @param <K> type of key
      * @param <V> type of ordered SPI class
      */
-    public static <K, V extends OrderedSPI<?>> void cacheServices(final Collection<K> types, final Class<V> orderedSPIClass, final Map<K, V> services) {
-        CACHED_SERVICES.put(orderedSPIClass, new CachedOrderedServices(types, services));
+    public static <K, V extends OrderedSPI<?>> void cacheServices(final OrderedServicesCacheKey cacheKey, final Map<K, V> services) {
```

Comment:
Is it necessary to add generic type `K` and `V`?

Reply:
> Is it necessary to add generic types `K` and `V`?

@terrymanu I will try to remove them.